### PR TITLE
[New Operator] StakeDAO + Yearn (curve pools)

### PR DESCRIPTION
**Pull request by Yashiru**
## Context

[StakeDAO](https://github.com/NestedFi/nested-core-lego/tree/feature/sc-54-operator-proposal-stakedao) and [Yearn](https://github.com/NestedFi/nested-core-lego/tree/feature/sc-80-operator-proposal-yearn-curve-vaults) branchs being very similar, it was necessary to merge them in order to factorize the code in common of these two branches.

The library `StakingLPVaultHelpers.sol` contain all the [StakeDAO](https://github.com/NestedFi/nested-core-lego/tree/feature/sc-54-operator-proposal-stakedao) and [Yearn](https://github.com/NestedFi/nested-core-lego/tree/feature/sc-80-operator-proposal-yearn-curve-vaults) factorized code. This file use `CurveHelpers.sol` library wich contain all the common Curve pool features.

To learn more about the initial PRs: 
- #118 
- #117 

## IStakingVault interface

The merging of these two branches resulted in the need to create a parent interface for `IStakeDaoStrategy.sol` and `IYearnVault.sol` in order to make them usable in the same way in the `StakingLPVaultHelpers.sol` library

### CODE REVIEW

File: contracts/libraries/CurveHelpers.sol [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    /// @param poolCoinAmount The amount of token in the Curve pool
+    /// @param signature The signature of the remove_liquidity_one_coin
+    ///                  function to be used to call to the Curve pool
+    /// @return success If the call to remove liquidity succeeded
+    function removeLiquidityOneCoin(
+        ICurvePool pool,
+        uint256 amount,
+        address outputToken,
+        uint256 poolCoinAmount,
+        string memory signature
```
Comment (maximebrugel) [inline contracts/libraries/CurveHelpers.sol:84]: We should use the selector (`bytes4` and not a `string`).

```solidity
// Convert string signature to bytes4 to use as argument
bytes4(keccak256(bytes("my_signature"))):
```

Then, you can use `abi.encodeWithSelector`

File: contracts/interfaces/external/ICurvePool/ICurvePool.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.11;
+
+/// @title Curve pool interface
+interface ICurvePool {
+    function token() external view returns (address);
+
+    function coins(uint256 index) external view returns (address);
+
+    function remove_liquidity_one_coin(
```
Comment (maximebrugel) [inline contracts/interfaces/external/ICurvePool/ICurvePool.sol:10]: `remove_liquidity_one_coin` function are never used with the Interface (only using the signature). We can remove them if we use directly the signature (literal).

Comment (maximebrugel) [inline contracts/interfaces/external/ICurvePool/ICurvePool.sol:10]: `remove_liquidity_one_coin` function are never used with the Interface (only using the signature). We can remove them if we use directly the signature (literal).

File: contracts/operators/StakeDAO/StakeDaoStrategyStorage.sol [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    mapping(address => CurvePool) public strategies;
+
+    /// @notice Add a StakeDAO strategy
+    /// @param strategy The strategy address
+    /// @param curvePool The underlying CurvePool (used to deposit)
+    function addStrategy(address strategy, CurvePool calldata curvePool) external onlyOwner {
+        require(strategy != address(0), "SDSS: INVALID_STRATEGY_ADDRESS");
+        require(curvePool.poolAddress != address(0), "SDSS: INVALID_POOL_ADDRESS");
+        require(strategies[strategy].poolAddress == address(0), "SDSS: ALREADY_EXISTENT_STRATEGY");
+        require(strategies[strategy].lpToken == address(0), "SDSS: ALREADY_EXISTENT_STRATEGY");
```
Comment (maximebrugel) [inline contracts/operators/StakeDAO/StakeDaoStrategyStorage.sol:37]: maybe using a different require message from the one before?

**Review by maximebrugel — APPROVED** [review]

File: contracts/libraries/StakingLPVaultHelpers.sol [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+import "./../Withdrawer.sol";
+import "../libraries/CurveHelpers.sol";
+import "./../libraries/ExchangeHelpers.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./../interfaces/external/ICurvePool/ICurvePool.sol";
+import "./../interfaces/external/ICurvePool/ICurvePool.sol";
+import "./../interfaces/external/ICurvePool/ICurvePoolETH.sol";
+import "../interfaces/external/IStakingVault/IStakingVault.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
```
[the remark this answers is not in the document]
Comment (Yashiru) [reply contracts/libraries/StakingLPVaultHelpers.sol:12]: I prefer to use the "cascade import" method to stay in line with the rest of the code.

File: contracts/operators/Yearn/YearnCurveVaultOperator.sol [hunk 5]
```diff
@@ -0,0 +1,284 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.11;
+
+import "./../../Withdrawer.sol";
+import "./YearnVaultStorage.sol";
+import "./../../libraries/ExchangeHelpers.sol";
+import "./../../interfaces/external/IWETH.sol";
+import "./../../libraries/CurveHelpers.sol";
+import "../../libraries/StakingLPVaultHelpers.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
```
[the remark this answers is not in the document]
Comment (Yashiru) [reply contracts/operators/Yearn/YearnCurveVaultOperator.sol:10]: I prefer to use the "cascade import" method to stay in line with the rest of the code.

File: contracts/operators/StakeDAO/StakeDaoStrategyStorage.sol (hunk 3 above)
Comment (Yashiru) [reply contracts/operators/StakeDAO/StakeDaoStrategyStorage.sol:37]: I also applied the same changes to `YearnVaultStorage.sol`.

[the remark this answers is not in the document]
Comment (Yashiru) [reply contracts/operators/StakeDAO/StakeDaoStrategyStorage.sol:37]: I also applied the same changes to `YearnVaultStorage.sol`.

File: contracts/libraries/CurveHelpers.sol [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /// @dev Get the arrays of obtained token and spent token
+    /// @param inputToken The token spent
+    /// @param inputTokenBalanceBefore The input token balance before
+    /// @param expectedInputAmount The expected amount of input token spent
+    /// @param outputToken The token obtained
+    /// @param outputTokenBalanceBefore The output token balance before
+    /// @param minAmountOut The minimum of output token expected
+    function getOutputAmounts(
```
[the remark this answers is not in the document]
Comment (obatirou) [reply contracts/libraries/CurveHelpers.sol]: [Issue linear created](https://linear.app/nested-finance/issue/SC-100/factorise-getoutputamount-in-libraries)
Closing comment

File: contracts/operators/StakeDAO/StakeDaoCurveStrategyOperator.sol [hunk 7]
```diff
@@ -0,0 +1,285 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.11;
+
+import "./../../Withdrawer.sol";
+import "./StakeDaoStrategyStorage.sol";
+import "../../libraries/CurveHelpers.sol";
+import "./../../interfaces/external/IWETH.sol";
+import "./../../libraries/ExchangeHelpers.sol";
+import "../../libraries/StakingLPVaultHelpers.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
```
[the remark this answers is not in the document]
Comment (obatirou) [reply contracts/operators/StakeDAO/StakeDaoCurveStrategyOperator.sol:11]: Ok !

**Review by maximebrugel — APPROVED** [review]

**Pull request merged**